### PR TITLE
Clean up diagnostics display

### PR DESCRIPTION
**Pull request by ahejlsberg**
Following some recent changes the output of the `-diagnostics` compiler option was no longer properly aligned and also somewhat confusing. This PR cleans it up. We now display:

* I/O read: Time spent reading from or searching for files in parse phase.
* I/O write: Time spent writing to files in emit phase.
* Parse time: Total time spent in parse phase (including I/O read).
* Bind time: Total time spent in bind phase.
* Check time: Total time spent in check phase.
* Emit time: Total time spent in emit phase (including I/O write).
* Total time: Sum of parse, bind, check, and emit.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -119,16 +122,19 @@ module ts {
...
+
         host = host || createCompilerHost(options);
 
+        var start = new Date().getTime();
         forEach(rootNames, name => processRootFile(name, false));
         if (!seenNoDefaultLib) {
             processRootFile(host.getDefaultLibFileName(options), true);
         }
-        verifyCompilerOptions();
+        programTime += new Date().getTime() - start;
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: Why is programTime set here?  It isn't measuring the time to verifyCompilerOptions it seems (and that's part of producing the program).

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -320,22 +320,16 @@ module ts {
...
+        checkTime = 0;
+        emitTime = 0;
 
         var program = createProgram(fileNames, compilerOptions, compilerHost);
-        var programTime = new Date().getTime() - start;
-
         var exitStatus = compileProgram();
 
-        var end = new Date().getTime() - start;
-        var compileTime = end - programTime;
```
Comment (CyrusNajmabadi) [inline src/compiler/tsc.ts]: Why was 'compileTime' removed?

File: src/compiler/tsc.ts [hunk 3]
```diff
@@ -356,19 +350,16 @@ module ts {
...
+            // I/O read time and processing time for triple-slash references and module imports, and the reported
+            // emit time includes I/O write time. We preserve this behavior so we can accurately compare times.
+            reportTimeStatistic("I/O read", ioReadTime);
+            reportTimeStatistic("I/O write", ioWriteTime);
             reportTimeStatistic("Parse time", programTime);
-            reportTimeStatistic("Bind time", ts.bindTime);
-            reportTimeStatistic("Check time", ts.checkTime);
-            reportTimeStatistic("Emit time", ts.emitTime);
-            
-            reportTimeStatistic("Parse time w/o IO", ts.parseTime);
```
Comment (CyrusNajmabadi) [inline src/compiler/tsc.ts]: We no longer report parse time without IO.  However, that's a useful thing to track.

File: src/compiler/program.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/program.ts]: Sure, easy enough to include.

**Comment by CyrusNajmabadi** [comment]
I'm not sure about this change.

I/O read: Time spent reading from or searching for files in parse phase.
This makes sense to me.

I/O write: Time spent writing to files in emit phase.
This makes sense as well.

Parse time: Total time spent in parse phase (including I/O read).
1) This makes sense wrt trying to keep things consistent with 1.3.  However, a better approach might be to instead tweak the 1.3 compiler to more properly break down its own numbers.

2) We've now removed the ability to just find out the cost of parsing.  You can do "parse time - io read time", but this assumes that depends on knowing internal implementation details of the system.  Previously the code just measured parse time and reported that.  It seems valuable to preserve that behavior.


Bind time: Total time spent in bind phase.
Check time: Total time spent in check phase.
Emit time: Total time spent in emit phase (including I/O write).
These seem ok.

Total time: Sum of parse, bind, check, and emit.
This seems wrong to me.  These are sub measurements of the entire compile process.  However, they're not the 'total' time.  There are still other things we may do while compiling.  The previous change reported the value exactly (and broke it down into the two main things we do in TSC: making the program, and the compiling it).  

The older approach gave accurate times about the total compile, and accurate numbers about the major subsystems.  The new one seems to depend a lot more on inferring some of these values from others, and also depends on knowing the internal implementation choices of the compiler.

**Comment by ahejlsberg** [comment]
We can't tweak the 1.3 compiler (or any of the ones that went before it). We need to stick with reporting the same numbers. With this PR the new output just adds two extra numbers but otherwise it is exactly the same (and grouped the same). I think that consistency is valuable.

Cost of raw parsing is easy, just subtract I/O read. Likewise, cost of raw emit is easy, just subtract I/O write. Not hard to do, better than reporting even more numbers.

Regarding total time, reported total used to be the sum of parse + bind + check + emit. Now it is again. Also, this revision removes the time it takes to output error messages to stdout which really isn't a meaningful thing to include.

Yes they're sub-measurements, so is everything else (because we don't know when JIT happens, etc.).

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**